### PR TITLE
Replaced the runtime installation of the SUSE userspace `fuse`

### PR DESCRIPTION
**Pull request by andreas-kupries**
Was reaching out to unknown machines, installing an unknown version with a combination of build- and runtime installation. Prevents air-gapped deployment.

Ref: https://trello.com/c/dOiHIcaP/847-5-update-nfs-volume-release-to-176

Installation of the `fuse` package moved into the stemcells, see PR #2251.

This PR modifies the patch for the `mapfs` pre-start script to insert a skip over the installation of the pre-packaged .deb archives when `fuse` support is already installed, instead of inserting the old full installation of `fuse` support via zypper and network.

----

Testing: Build the branch and start the cluster. Run the test suites.

### CODE REVIEW

File: src/scf-release/jobs/install-suse-fuse/templates/pre-start.erb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash -e
+##
+set -o xtrace
+set -o errexit -o nounset
+# This installs the FUSE userspace support.
+# The required files are found in a tarball
+# provided by the `suse-fuse` package.
```
Comment (andreas-kupries) [inline src/scf-release/jobs/install-suse-fuse/templates/pre-start.erb]: No. It is the original SUSE rpm archive.

File: src/scf-release/packages/suse-fuse/packaging [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+SRC=/var/cache/zypp/packages
+
+# Clear the rpm cache. We want to see later only new files.
+zypper clean
+
+# Get fuse rpm and dependencies
+zypper --non-interactive --quiet install --download-only fuse
+
+# Get and copy the content of the rpm cache into the package-to-be.
+cp $(find /var/cache/ -type f -name '*.rpm') ${DST}/
```
Comment (andreas-kupries) [inline src/scf-release/packages/suse-fuse/packaging]: Oops. The directory should be `${SRC}`

File: src/scf-release/packages/suse-fuse/packaging [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+# Show what we packaged.
+echo ................................. packaged
+find -L $DST -type f
+
+# Force error and log
```
Comment (andreas-kupries) [inline src/scf-release/packages/suse-fuse/packaging]: Nonsense comment. Remove.

File: src/scf-release/packages/suse-fuse/packaging [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+set -e -x
+
+DST=${BOSH_INSTALL_TARGET}
+SRC=/var/cache/zypp/packages
+
+# Clear the rpm cache. We want to see later only new files.
+zypper clean
+
+# Get fuse rpm and dependencies
+zypper --non-interactive --quiet install --download-only fuse
```
Comment (andreas-kupries) [inline src/scf-release/packages/suse-fuse/packaging]: Jenkins fails here with `16:59:57 compilation-suse-fuse > No provider of 'fuse' found.`
Have to check what set of repos are known here.
Difference to Vagrant which was ok.

Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Should remove the `--quiet`

File: src/scf-release/jobs/install-suse-fuse/spec [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+name: install-suse-fuse
+description: Install the SUSE userspace FUSE support.
+
+templates:
+  pre-start.erb: bin/pre-start
+
+packages:
+  - suse-fuse
```
Comment (jandubois) [inline src/scf-release/jobs/install-suse-fuse/spec]: No need to indent list elements:

```
packages:
- suse-fuse
```

File: src/scf-release/packages/suse-fuse/packaging [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+set -e -x
+
+DST=${BOSH_INSTALL_TARGET}
+SRC=/var/cache/zypp/packages
+
+# Clear the rpm cache. We want to see later only new files.
+zypper clean
+
+# Get fuse rpm and dependencies
+zypper --non-interactive install --download-only fuse
```
Comment (jandubois) [inline src/scf-release/packages/suse-fuse/packaging]: I'm a bit of two minds here:

1. I think we should specify the precise package version here. A bosh release should always compile to the same bits and not incorporate whatever is latest at the time it is compiled. The Ubuntu fuse packages are embedded in the bosh release and don't change when Ubuntu updates the package.

2. Hard-coding the fuse version is a maintenance headache. We would have to update the bosh release whenever there is a CVE to this package, and we don't have a mechanism to notify us. However, the situation is only accidentally better by not hard-coding the version: we only rebuild when there is a new release of the full product.

In the end this is an issue produced by upstream. It is simply wrong to have stemcell-specific bits inside a bosh release. ☹️

File: src/scf-release/packages/suse-fuse/spec [hunk 7]
```diff
@@ -0,0 +1,3 @@
+---
+name: suse-fuse
+files:
```
Comment (jandubois) [inline src/scf-release/packages/suse-fuse/spec]: I would prefer

```
files: []
```
to indicate that this is an empty list and not just `nil`, but I guess it doesn't make any difference. However, it also makes it obvious that there are not missing lines at the end that got accidentally deleted...

File: src/scf-release/jobs/install-suse-fuse/templates/pre-start.erb (hunk 1 above)
Comment (andreas-kupries) [reply src/scf-release/jobs/install-suse-fuse/templates/pre-start.erb]: Done with commit 838d12fad5710bd0aaa9d0a107f1fe1b0688a90f

File: src/scf-release/packages/suse-fuse/packaging (hunk 2 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Done with commit 838d12fad5710bd0aaa9d0a107f1fe1b0688a90f, then superseded in commit 571a972ea60268849419855b642ed8be94ef47fa

File: src/scf-release/packages/suse-fuse/packaging (hunk 3 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Done with commit 838d12fad5710bd0aaa9d0a107f1fe1b0688a90f

File: src/scf-release/packages/suse-fuse/packaging (hunk 4 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Added diagnostics with commit 571a972ea60268849419855b642ed8be94ef47fa.
Currently suspect:

- Vagrant = opensuse stemcell, has predefined repos to look at
- Jenkins = sle stemcell, no repos, or other

File: src/scf-release/jobs/install-suse-fuse/spec (hunk 5 above)
Comment (andreas-kupries) [reply src/scf-release/jobs/install-suse-fuse/spec]: Done with commit f5ad84cb4a10919001addd629b9cafd7c85bd1a1

File: src/scf-release/packages/suse-fuse/spec (hunk 7 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/spec]: Done with commit 795b58f9da96e0e6342529f0c244d37648fd4a6d

File: src/scf-release/packages/suse-fuse/packaging (hunk 6 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: I can see both sides and am also two-minded. No decision yet. Lets see if in the end the fuse version I get differs for opensuse and sle setups.

File: src/scf-release/packages/suse-fuse/packaging (hunk 4 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Confirmed. OpenSuse stemcell knows 3 repos, opensuse distribution + update, and stemcell repo.
SLE stemcell only knows the stemcell repo.

Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: By adding the relevant repos I I will also get the same fuse for both I guess. Using a fixed version should therefore be no problem.

Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Ref commit c6471b01a4c78e864a42e30e27e0a898eba062b2, trialing

Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: Past the build stage now with commits 23cd7f909e4a886a1a3337403a91cabe999f79c6 and 8f344a3cb8b112b3bbaad9ebb76b03dc866a5a6e fixing a few zypper options.

File: src/scf-release/packages/suse-fuse/packaging (hunk 6 above)
Comment (andreas-kupries) [reply src/scf-release/packages/suse-fuse/packaging]: outdated with the move of fuse install to stemcells.

File: container-host-files/etc/scf/config/scripts/patches/mapfs_pre_start.sh [hunk 8]
```diff
@@ -2,32 +2,23 @@
...
   exit 0
 fi
 
-patch -d "$PATCH_DIR" --force -p1 <<'PATCH'
-From c46738e970913e75d04f0654df6033c6e0ad7d61 Mon Sep 17 00:00:00 2001
-From: Mark Yen <mark.yen@suse.com>
-Date: Wed, 30 Jan 2019 13:51:36 -0800
-Subject: [PATCH] mapfs: install: Support for SUSE
+# This adds support for SUSE-like systems to install fuse:
+# - Detect installed fuse support and skip installation of pre-packaged (ubuntu) debs
```
Comment (f0rmiga) [inline container-host-files/etc/scf/config/scripts/patches/mapfs_pre_start.sh]: [nit] This is the only point without punctuation.

**Pull request merged**